l1/ecl_l1_pos_controller.cpp

```diff
@@ -70,20 +70,40 @@ float ECL_L1_Pos_Controller::switch_distance(float wp_radius)
 	return math::min(wp_radius, _L1_distance);
 }
 
-void
-ECL_L1_Pos_Controller::navigate_waypoints(const Vector2f &vector_A, const Vector2f &vector_B,
-		const Vector2f &vector_curr_position, const Vector2f &ground_speed_vector)
+float ECL_L1_Pos_Controller::airspeed_incr(float airspeed, const float airsp_ref, const float airsp_max,
+		const float wind_speed, float min_ground_speed)
+{
+	/* following [3] */
+	// NOTE: uses current "lambda" angle, this function should only be called *after any other navigation function
+
+	const float airsp_incr_max = math::max(airsp_max - airsp_ref, 0.0f);
+
+	airspeed = math::max(airspeed, 0.1f);
+	min_ground_speed = math::max(min_ground_speed, 0.0f);
+
+	const float wind_ratio = (wind_speed + min_ground_speed) / airspeed;
```

Comment:
We will need to do a sensitivity analysis to determine the effect of wind speed estimation errors on this method. We may have to add some additional protection using the dot product of the X body axis projected onto earth axes and the velocity vector to determine when forwards speed has dropped below a threshold and increment the airspeed demand accordingly.

Reply:
Totally. I've done some minor playing around with the effects of wrong wind estimates (mainly cases where the airspeed sensor is not properly calibrated and either over or under estimates) -- small errors (order of 1-2m/s) don't effect things much -- but it has been on my list to do a proper analysis (just haven't gotten around to it). However I guess I probably should have also made it clear that this functionality **relies** on a proper wind estimate. So if e.g. the airspeed sensor is not calibrated, or sideslip fusion has some significant uncertainty, it will not work as nicely -- however, if the wind estimates are good enough, it is much more efficient than the ground speed undershoot solution. Nevertheless, I agree with you that it is probably a good idea to put back in a fallback solution to some form of safe guard with the x-body - gsp vector dot product especially when the wind estimate e.g. goes invalid or has high uncertainty.

Something to note: the logic is currently not used when the wind estimate is invalid (defaults to ground speed and "zero" wind speed as input args) -- this was one of the "open questions" checkboxes I left on the PR, that we would need to think what to do when this case happens -- so I guess again this points towards a defaulted ground speed undershoot. I'll look into how to put it back in and make it play nice with the other logic in the next day or so.